File: 0049-group-anagrams/0049-group-anagrams.py
'''
T: O(N *MlogM)
S: O(N)
'''
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        if len(strs)==0:
            return [[""]]
        
        output = {}
        
        for word in strs: # O(n)
            sortedWord = ''.join(sorted(word)) # n*log(n)
            if sortedWord in output:
                output[sortedWord].append(word)
            else:
                output[sortedWord] = [word]
                
        return output.values()

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        /*
            Algorithms & DS: HashMap
            Time: O(N) , Space: O(N)
        
        */
        HashMap<String, List<String>> map = new HashMap<>();
        List<List<String>> result = new ArrayList<>();

        for(int i = 0; i < strs.length; i++){
            String val = strs[i];
            char[] sorted = val.toCharArray();
            Arrays.sort(sorted);
            String key = new String(sorted);

            if(!map.containsKey(key)) map.put(key, new ArrayList<>());
            map.get(key).add(val);
        }

        for(String key : map.keySet()){
            result.add(map.get(key));
        }

        return result;
    }
}